### PR TITLE
Fix error in calculating bounding box 

### PR DESCRIPTION
**Pull request by johanzhu**

### CODE REVIEW

File: example/outline.ts [hunk 1]
```diff
@@ -74,6 +74,7 @@ export default class BoundingBoxLine extends MeshRenderer {
 
     const rendererCount = renderers.length;
     if (rendererCount > 0) {
+      if (!renderers[0].bounds) return;
```
Comment (singlecoder) [inline example/outline.ts]: This line of judgment is unnecessary; all renderers will have bounds

File: src/SpineAnimationRenderer.ts [hunk 2]
```diff
@@ -104,6 +133,12 @@ export class SpineAnimationRenderer extends Renderer {
   /** @internal */
   @ignoreClone
   _resource: SpineResource;
+  /** @internal */
+  @ignoreClone
+  _spineBounds = new BoundingBox(
```
Comment (singlecoder) [inline src/SpineAnimationRenderer.ts]: why not use _localBounds

File: src/SpineAnimationRenderer.ts [hunk 3]
```diff
@@ -153,32 +188,42 @@ export class SpineAnimationRenderer extends Renderer {
...
+        this._setDirtyFlagFalse(SpineAnimationUpdateFlags.AnimationVolume);
       }
     }
+    SpineAnimationRenderer._spineGenerator.buildPrimitive(
+      this._skeleton,
+      this,
+      shouldUpdateBounds
+    );
+    if (shouldUpdateBounds) {
+      this._updateBounds(this.bounds);
```
Comment (singlecoder) [inline src/SpineAnimationRenderer.ts]: Should use this._bounds

File: example/outline.ts (hunk 1 above)
Comment (johanzhu) [reply example/outline.ts]: done

File: src/SpineAnimationRenderer.ts (hunk 2 above)
Comment (johanzhu) [reply src/SpineAnimationRenderer.ts]: done

File: src/SpineAnimationRenderer.ts [hunk 4]
```diff
@@ -175,10 +216,6 @@ export class SpineAnimationRenderer extends Renderer {
    */
   // @ts-ignore
   override _render(context: any): void {
-    if (this._needsInitialize) {
-      this._initialize();
-      this._needsInitialize = false;
-    }
     const { _primitive, _subPrimitives } = this;
     const { _materials: materials, _engine: engine } = this;
     // @ts-ignore
```
Comment (singlecoder) [inline src/SpineAnimationRenderer.ts:203]: Put together to retrieve values

File: src/SpineAnimationRenderer.ts [hunk 5]
```diff
@@ -269,6 +298,7 @@ export class SpineAnimationRenderer extends Renderer {
       _primitive.destroy();
     }
     this._clearMaterialCache();
+    this._resource && this._resource.destroy();
     this._primitive = null;
```
Comment (singlecoder) [inline src/SpineAnimationRenderer.ts:274]: _onDestroy should be more concise, like:
```typescript
override _onDestroy(): void {
  this._clearMaterialCache();
  this._subPrimitives.length = 0;
  this._primitive && this._primitive.destroy();
  this._resource && this._resource.destroy();
  this._primitive = null;
  this._resource = null;
  this._skeleton = null;
  this._state = null;
  super._onDestroy();
}
```

File: src/SpineAnimationRenderer.ts [hunk 6]
```diff
@@ -333,40 +371,9 @@ export class SpineAnimationRenderer extends Renderer {
...
-    if (!entry.loop) {
-      this._setDirtyFlagFalse(SpineAnimationUpdateFlags.AnimationVolume)
-    }
-  }
-
   private _clearMaterialCache(): void {
     this._materials.forEach((item) => {
-      const texture = item.shaderData.getTexture('material_SpineTexture');
+      const texture = item.shaderData.getTexture("material_SpineTexture");
       const blendMode = getBlendMode(item);
```
Comment (singlecoder) [inline src/SpineAnimationRenderer.ts]: 1、Cache SpineAnimationRenderer._materialCache outside the loop
2、rename item to material may be more intuitive

File: src/SpineAnimationRenderer.ts [hunk 7]
```diff
@@ -153,20 +188,26 @@ export class SpineAnimationRenderer extends Renderer {
...
+    skeleton.updateWorldTransform(Physics.update);
+    SpineAnimationRenderer._spineGenerator.buildPrimitive(
+      this._skeleton,
+      this,
+      shouldUpdateBounds
+    );
+    if (shouldUpdateBounds) {
+      // @ts-ignore
+      this._updateBounds(this._bounds);
     }
```
Comment (singlecoder) [inline src/SpineAnimationRenderer.ts]: suggestion
```typescript
shouldUpdateBounds && this._updateBounds(this._bounds);
```

File: src/SpineGenerator.ts [hunk 8]
```diff
@@ -372,15 +362,25 @@ export class SpineGenerator {
...
+  ): Material {
     const material = SpineAnimationRenderer._getDefaultMaterial(engine);
     material.shaderData.setTexture("material_SpineTexture", texture);
     setBlendMode(material, blendMode, premultipliedAlpha);
     return material;
   }
 
-  private _expandByPoint(x: number, y: number, z: number) {
-    const { bounds: { min, max } } = SpineGenerator;
+  private _expandByPoint(
```
Comment (singlecoder) [inline src/SpineGenerator.ts]: suggestion
```typescript
private _expandByPoint(
  x: number,
  y: number,
  z: number,
  renderer: SpineAnimationRenderer
) {
  const { min, max } = renderer._localBounds;
  min.set(Math.min(min.x, x), Math.min(min.y, y), Math.min(min.z, z));
  max.set(Math.max(max.x, x), Math.max(max.y, y), Math.max(max.z, z));
}
```

Comment (johanzhu) [reply src/SpineGenerator.ts]: done

File: src/SpineAnimationRenderer.ts (hunk 7 above)
Comment (johanzhu) [reply src/SpineAnimationRenderer.ts]: done

File: src/SpineAnimationRenderer.ts (hunk 6 above)
Comment (johanzhu) [reply src/SpineAnimationRenderer.ts]: done

File: src/SpineAnimationRenderer.ts (hunk 5 above)
Comment (johanzhu) [reply src/SpineAnimationRenderer.ts:274]: done

File: src/SpineAnimationRenderer.ts (hunk 4 above)
Comment (johanzhu) [reply src/SpineAnimationRenderer.ts:203]: done

File: src/SpineGenerator.ts [hunk 9]
```diff
@@ -192,26 +177,25 @@ export class SpineGenerator {
             uvs,
             finalColor,
             dark,
```
Comment (singlecoder) [inline src/SpineGenerator.ts]: dark not need cache, just use once

File: src/SpineGenerator.ts [hunk 10]
```diff
@@ -348,15 +323,20 @@ export class SpineGenerator {
       const key = `${subTexture.instanceId}_${blendMode}`;
       let material = SpineAnimationRenderer._materialCache.get(key);
       if (!material) {
-        material = this._createMaterialForTexture(subTexture, engine, blendMode, premultipliedAlpha);
+        material = this._createMaterialForTexture(
```
Comment (singlecoder) [inline src/SpineGenerator.ts:319]:
```typescript
if (curLen < lastLen) {
    for (let i = curLen; i < lastLen; i++) {
      const item = _subPrimitives[i];
      subPrimitivePool.return(item);
    }
  }
```

The if condition in the above code is not needed

```typescript
let subTexture = texture.texture;
if (_separateSlotTextureMap.has(slotName)) {
  subTexture = _separateSlotTextureMap.get(slotName);
}
```

should opt

```typescript
let subTexture = _separateSlotTextureMap.get(slotName) || texture.texture;
```

File: src/SpineGenerator.ts [hunk 11]
```diff
@@ -348,15 +323,20 @@ export class SpineGenerator {
       const key = `${subTexture.instanceId}_${blendMode}`;
       let material = SpineAnimationRenderer._materialCache.get(key);
```
Comment (singlecoder) [inline src/SpineGenerator.ts]: _materialCache should cache out loop

File: src/SpineGenerator.ts (hunk 10 above)
Comment (johanzhu) [reply src/SpineGenerator.ts:319]: done

File: src/SpineGenerator.ts (hunk 11 above)
Comment (johanzhu) [reply src/SpineGenerator.ts]: done

File: src/SpineGenerator.ts (hunk 9 above)
Comment (johanzhu) [reply src/SpineGenerator.ts]: done

**Comment by singlecoder** [comment]
+1

**Review by singlecoder — APPROVED** [review]

File: src/SpineAnimationRenderer.ts [hunk 12]
```diff
@@ -70,12 +91,20 @@ export class SpineAnimationRenderer extends Renderer {
   @assignmentClone
   premultipliedAlpha = false;
 
+  /**
+   * Updates the bounding box every frame when enabled.
+   * Otherwise, it is calculated only during animation initialization.
+   */
+  @assignmentClone
+  updateBoundsPerFrame = false;
```
Comment (GuoLei1990) [inline src/SpineAnimationRenderer.ts]: Delete this

File: src/SpineAnimationRenderer.ts [hunk 13]
```diff
@@ -153,34 +188,38 @@ export class SpineAnimationRenderer extends Renderer {
...
+    state.apply(skeleton);
+    skeleton.update(delta);
+    skeleton.updateWorldTransform(Physics.update);
+    SpineAnimationRenderer._spineGenerator.buildPrimitive(
+      this._skeleton,
+      this,
+      shouldUpdateBounds
+    );
+    // @ts-ignore
+    shouldUpdateBounds && this._updateBounds(this._bounds);
```
Comment (GuoLei1990) [inline src/SpineAnimationRenderer.ts]: Shouldn't depend on update, should get to compute 

You can check `MeshRenderer` _updateBounds logic

File: src/SpineAnimationRenderer.ts [hunk 14]
```diff
@@ -263,12 +294,10 @@ export class SpineAnimationRenderer extends Renderer {
...
   override _onDestroy(): void {
-    const { _primitive, _subPrimitives } = this;
-    _subPrimitives.length = 0;
-    if (_primitive) {
-      _primitive.destroy();
-    }
     this._clearMaterialCache();
+    this._subPrimitives.length = 0;
+    this._primitive && this._primitive.destroy();
+    this._resource && this._resource.destroy();
```
Comment (GuoLei1990) [inline src/SpineAnimationRenderer.ts]: No! SpineResource is refer resource, You shouldn't do this

File: src/SpineAnimationRenderer.ts [hunk 15]
```diff
@@ -333,51 +370,23 @@ export class SpineAnimationRenderer extends Renderer {
...
+    const { _materials: materials } = this;
+    for (let i = 0, len = materials.length; i < len; i += 1) {
+      const material = materials[i];
+      const texture = material.shaderData.getTexture("material_SpineTexture");
+      const blendMode = getBlendMode(material);
       const key = `${texture.instanceId}_${blendMode}`;
-      SpineAnimationRenderer._materialCache.delete(key);
-    });
+      materialCache.delete(key);
+    }
```
Comment (GuoLei1990) [inline src/SpineAnimationRenderer.ts:341]: Very unreasonable material design！

File: src/SpineAnimationRenderer.ts (hunk 12 above)
Comment (johanzhu) [reply src/SpineAnimationRenderer.ts]: done

File: src/SpineAnimationRenderer.ts (hunk 13 above)
Comment (johanzhu) [reply src/SpineAnimationRenderer.ts]: done

File: src/SpineAnimationRenderer.ts (hunk 14 above)
Comment (johanzhu) [reply src/SpineAnimationRenderer.ts]: done

File: src/SpineAnimationRenderer.ts (hunk 15 above)
Comment (johanzhu) [reply src/SpineAnimationRenderer.ts:341]: The modification of materials involves the entire design of custom materials for Spine, which requires significant changes. This PR will not address this point for now. Really thank you for your understanding.

File: src/SpineAnimationRenderer.ts [hunk 16]
```diff
@@ -153,34 +181,31 @@ export class SpineAnimationRenderer extends Renderer {
...
-      if (this._isContainDirtyFlag(SpineAnimationUpdateFlags.InitialVolume)) {
-        this._onWorldVolumeChanged();
-        this._setDirtyFlagFalse(SpineAnimationUpdateFlags.InitialVolume);
-      }
-      if (this._isContainDirtyFlag(SpineAnimationUpdateFlags.AnimationVolume)) {
-        this._calculateGeneratorBounds(this.bounds);
-      }
+    const { _state: state, _skeleton: skeleton } = this;
+    if (!state || !skeleton) return;
+    if (this._needsInitialize) {
```
Comment (GuoLei1990) [inline src/SpineAnimationRenderer.ts]: emmm!

File: src/SpineGenerator.ts [hunk 17]
```diff
@@ -220,58 +200,53 @@ export class SpineGenerator {
...
+          _vertices[i++] = x;
+          _vertices[i++] = y;
+          _vertices[i++] = z;
+          _vertices[i++] = finalVertices[j++];
+          _vertices[i++] = finalVertices[j++];
+          _vertices[i++] = finalVertices[j++];
+          _vertices[i++] = finalVertices[j++];
+          _vertices[i++] = finalVertices[j++];
+          _vertices[i++] = finalVertices[j++];
+          this._expandByPoint(x, y, z, renderer);
```
Comment (GuoLei1990) [inline src/SpineGenerator.ts]: `renderer`->localBounds

File: src/SpineAnimationRenderer.ts (hunk 16 above)
Comment (johanzhu) [reply src/SpineAnimationRenderer.ts]: done

File: src/SpineGenerator.ts (hunk 17 above)
Comment (johanzhu) [reply src/SpineGenerator.ts]: done

**Review by GuoLei1990 — APPROVED** [review]

**Pull request merged**